config/dependency.go

```diff
@@ -114,6 +115,86 @@ func getTerragruntOutput(dependencyConfig Dependency, terragruntOptions *options
+// - If the dependency block indicates a default_outputs attribute, this will return that.
+// - If the dependency block does NOT indicate a default_outputs attribute, this will return an error.
+func getTerragruntOutputIfAppliedElseConfiguredDefault(dependencyConfig Dependency, terragruntOptions *options.TerragruntOptions) (*cty.Value, error) {
+	isApplied, err := isTerragruntApplied(dependencyConfig, terragruntOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	if isApplied {
+		return getTerragruntOutput(dependencyConfig, terragruntOptions)
+	}
+
+	// When the module is not applied, check if there is a default outputs value to return. If yes, return that. Else,
+	// return error.
+	targetConfig := getCleanedTargetConfigPath(dependencyConfig, terragruntOptions)
+	currentConfig := terragruntOptions.TerragruntConfigPath
+	if shouldReturnDefaultOutputs(dependencyConfig, terragruntOptions) {
+		util.Debugf(
+			terragruntOptions.Logger,
+			"WARNING: config %s is a dependency of %s that hasn't been applied, but default outputs provided and returning those in dependency output.",
+			targetConfig,
+			currentConfig,
+		)
+		return dependencyConfig.DefaultOutputs, nil
+	}
+
+	err = TerragruntOutputTargetNotApplied{
+		targetConfig:  targetConfig,
+		currentConfig: currentConfig,
+	}
+	return nil, err
+}
+
+// We should only return default outputs if the default_outputs attribute is set, and if we are running one of the
+// allowed commands when `default_outputs_allowed_terraform_commands` is set as well.
+func shouldReturnDefaultOutputs(dependencyConfig Dependency, terragruntOptions *options.TerragruntOptions) bool {
+	defaultOutputsSet := dependencyConfig.DefaultOutputs != nil
+	allowedCommand :=
+		dependencyConfig.DefaultOutputsAllowedTerraformCommands == nil ||
+			len(*dependencyConfig.DefaultOutputsAllowedTerraformCommands) == 0 ||
+			util.ListContainsElement(*dependencyConfig.DefaultOutputsAllowedTerraformCommands, terragruntOptions.TerraformCommand)
+	return defaultOutputsSet && allowedCommand
+}
+
+// Return whether or not the target config has been applied. Returns false when the target config hasn't been applied,
+// and true when it has.
+func isTerragruntApplied(dependencyConfig Dependency, terragruntOptions *options.TerragruntOptions) (bool, error) {
+	// target config check: make sure the target config exists
+	targetConfig := getCleanedTargetConfigPath(dependencyConfig, terragruntOptions)
+	if !util.FileExists(targetConfig) {
+		return false, errors.WithStackTrace(DependencyConfigNotFound{Path: targetConfig})
+	}
+
+	// Check if the target terraform module (managed by terragrunt target config) has been applied.
+	// When a module hasn't been applied yet, there will be no state object, so `terragrunt show` will indicate that.
+	stdout, err := runTerragruntShow(terragruntOptions, targetConfig)
+	stdout = strings.TrimSpace(stdout)
+	isApplied := stdout != "No state."
```

Comment:
The downside to calling `show` is (a) it's an extra Terragrunt command to run, making this whole outputs thing a bit slower, (b) we are relying an an unofficial API.

Does the `terraform output` command perhaps return a specific exit code if there are no outputs or the module was never applied?

Reply:
> Does the terraform output command perhaps return a specific exit code if there are no outputs or the module was never applied?

It returns `{}` with no error, but that can be easily checked so added logic to check for this.

Reply:
UPDATE: this no longer uses show and just checks for empty object.

Reply:
Hm, so this is interesting. If you run `terraform output` on an un-applied module, you get an error, with an exit code of 1:

```
$ terraform output
The state file either has no outputs defined, or all the defined
outputs are empty. Please define an output in your configuration
with the `output` keyword and run `terraform refresh` for it to
become available. If you are using interpolation, please verify
the interpolated value is not empty. You can use the 
`terraform console` command to assist.
```

But if you run `terraform output -json`, you get no error and an exit code of 0:

```
$ terraform output -json
{}
```

That... seems problematic.

Reply:
Yup...

In fact,

> If you run terraform output on an un-applied module, you get an error

also returns an error if you have either of the following conditions as well:
- You have no `output` blocks
- All the output values are `null`.